src/theory/sets/theory_sets_type_enumerator.cpp

```diff
@@ -0,0 +1,132 @@
+/*! \file theory_sets_type_enumerator.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Kshitij Bansal, Tim King, Andrew Reynolds, Mudathir Mahgoub
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief set enumerator implementation
+ **
+ ** set enumerator implementation
+ **/
+
+#include "theory/sets/theory_sets_type_enumerator.h"
+
+namespace CVC4 {
+namespace theory {
+namespace sets {
+
+SetEnumerator::SetEnumerator(TypeNode type, TypeEnumeratorProperties* tep)
+    : TypeEnumeratorBase<SetEnumerator>(type),
+      d_nodeManager(NodeManager::currentNM()),
+      d_elementEnumerator(type.getSetElementType(), tep),
+      d_isFinished(false),
+      d_currentSetIndex(0),
+      d_currentSet()
+{
+  d_currentSet = d_nodeManager->mkConst(EmptySet(type.toType()));
+}
+
+SetEnumerator::SetEnumerator(const SetEnumerator& enumerator)
+    : TypeEnumeratorBase<SetEnumerator>(enumerator.getType()),
+      d_nodeManager(enumerator.d_nodeManager),
+      d_elementEnumerator(enumerator.d_elementEnumerator),
+      d_isFinished(enumerator.d_isFinished),
+      d_currentSetIndex(enumerator.d_currentSetIndex),
+      d_currentSet(enumerator.d_currentSet)
+{
+}
+
+SetEnumerator::~SetEnumerator() {}
+
+Node SetEnumerator::operator*()
+{
+  if (d_isFinished)
+  {
+    throw NoMoreValuesException(getType());
+  }
+
+  Trace("set-type-enum") << "SetEnumerator::operator* d_currentSet = "
+                         << d_currentSet << std::endl;
+
+  return d_currentSet;
+}
+
+SetEnumerator& SetEnumerator::operator++()
+{
+  if (d_isFinished)
+  {
+    Trace("set-type-enum") << "SetEnumerator::operator++ finished!"
+                           << std::endl;
+    Trace("set-type-enum") << "SetEnumerator::operator++ d_currentSet = "
+                           << d_currentSet << std::endl;
+    return *this;
+  }
+
+  d_currentSetIndex++;
+
+  // if the index is a power of 2, get a new element from d_elementEnumerator
+  if (d_currentSetIndex == (unsigned)(1 << d_elementsSoFar.size()))
+  {
+    // if there are no more values from d_elementEnumerator, set d_isFinished
+    // to true
+    if (d_elementEnumerator.isFinished())
+    {
+      d_isFinished = true;
+
+      Trace("set-type-enum")
+          << "SetEnumerator::operator++ finished!" << std::endl;
+      Trace("set-type-enum")
+          << "SetEnumerator::operator++ d_currentSet = " << d_currentSet
+          << std::endl;
+      return *this;
+    }
+
+    // get a new element and return it as a singleton set
+    Node element = *d_elementEnumerator;
+    d_elementsSoFar.push_back(element);
+    d_currentSet = d_nodeManager->mkNode(Kind::SINGLETON, element);
```

Comment:
```suggestion
    d_currentSet = d_nodeManager->mkNode(kind::SINGLETON, element);
```

Reply:
Done. wondering what is the difference between `Kind` and `kind`?